honeybadgermpc/protocols/avid.py

```diff
@@ -0,0 +1,216 @@
+                # send each person the column of stripes
+                self.send(i, (sid, 'VAL', roothash_list,
+                              branch_list, stripes_list_per_party[i]))
+
+        # counters and variables for msg
+        echo_set = set()
+        ready_set = set()
+        ready_sent = False
+        from_leader = None
+        # internal storage
+        my_stripes = [None] * self.input_size
+        my_roothash_list = [None] * self.input_size
+
+        while True:  # main recv loop
+            sender, msg = await self.recv()
+            if msg[1] == 'VAL' and from_leader is None:
+                # Validation
+                (_, _, roothash_list, branch_list, stripes_for_each) = msg
+                if sender != self.leader:
+                    logger.warning(
+                        f"[{pid}] VAL message from other than leader: {sender}")
+                    continue
+
+                # merkle tree verification
+                validation_fail_flag = False
+                for i in range(len(stripes_for_each)):
+                    # verify each entry in the stripes
+                    try:
+                        assert merkle_verify(self.n, stripes_for_each[i],
+                                             roothash_list[i], branch_list[i], pid)
+                    except Exception as e:
+                        logger.info(f"[{pid}]Failed to validate VAL message: {e}")
+                        validation_fail_flag = True
+                        break
+                if validation_fail_flag:
+                    continue
+
+                # Update and store the stripes & roothashes for retrieve
+                from_leader = pid
+                my_stripes = stripes_for_each
+                my_roothash_list = roothash_list
+
+                # echo others
+                broadcast((sid, 'ECHO'))
+
+            elif msg[1] == 'ECHO':
+                # Validation
+                if sender in echo_set:
+                    logger.warning("[{pid}] Redundant ECHO")
+                    continue
+
+                # Update
+                echo_set.add(sender)
+
+                if len(echo_set) >= echo_threshold and not ready_sent:
+                    ready_sent = True
+                    broadcast((sid, 'READY'))
+
+            elif msg[1] == 'READY':
+                # Validation
+                if sender in ready_set:
+                    logger.warning("[{pid}] Redundant READY")
+                    continue
+
+                # Update
+                ready_set.add(sender)
+
+                # Amplify ready messages
+                if len(ready_set) >= ready_threshold and not ready_sent:
+                    ready_sent = True
+                    broadcast((sid, 'READY'))
+
+                if len(ready_set) >= output_threshold and len(echo_set) >= k:
+                    # update ok queue indicating ready for retrieve
+                    self.ok_queue.put_nowait(True)
```

Comment:
Can you get multiple messages in the `ok` queue corresponding to different things? Is it possible that you add an `ok` message in the queue `n` times and then call retrieve `k` `k < n` times and then lets say someone else comes and `retrieve` sees an `ok` in the queue and returns the value even though it does not probably correspond to this retrieve.


What I am trying to get at is that can multiple OK messages cause any problem for retrieve?

Reply:
I think I'll update it to future